internal/upstreamldap/upstreamldap.go

```diff
@@ -170,71 +170,80 @@ func (p *Provider) GetConfig() ProviderConfig {
 		p.traceRefreshFailure(t, err)
-		return err
+		return nil, err
 	}
 
 	// if any more or less than one entry, error.
 	// we don't need to worry about logging this because we know it's a dn.
 	if len(searchResult.Entries) != 1 {
-		return fmt.Errorf(`searching for user %q resulted in %d search results, but expected 1 result`,
+		return nil, fmt.Errorf(`searching for user %q resulted in %d search results, but expected 1 result`,
 			userDN, len(searchResult.Entries),
 		)
 	}
 
 	userEntry := searchResult.Entries[0]
 	if len(userEntry.DN) == 0 {
-		return fmt.Errorf(`searching for user with original DN %q resulted in search result without DN`, userDN)
+		return nil, fmt.Errorf(`searching for user with original DN %q resulted in search result without DN`, userDN)
 	}
 
 	newUsername, err := p.getSearchResultAttributeValue(p.c.UserSearch.UsernameAttribute, userEntry, userDN)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	if newUsername != storedRefreshAttributes.Username {
-		return fmt.Errorf(`searching for user %q returned a different username than the previous value. expected: %q, actual: %q`,
+		return nil, fmt.Errorf(`searching for user %q returned a different username than the previous value. expected: %q, actual: %q`,
 			userDN, storedRefreshAttributes.Username, newUsername,
 		)
 	}
 
 	newUID, err := p.getSearchResultAttributeRawValueEncoded(p.c.UserSearch.UIDAttribute, userEntry, userDN)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	newSubject := downstreamsession.DownstreamLDAPSubject(newUID, *p.GetURL())
 	if newSubject != storedRefreshAttributes.Subject {
-		return fmt.Errorf(`searching for user %q produced a different subject than the previous value. expected: %q, actual: %q`, userDN, storedRefreshAttributes.Subject, newSubject)
+		return nil, fmt.Errorf(`searching for user %q produced a different subject than the previous value. expected: %q, actual: %q`, userDN, storedRefreshAttributes.Subject, newSubject)
 	}
 	for attribute, validateFunc := range p.c.RefreshAttributeChecks {
 		err = validateFunc(userEntry, storedRefreshAttributes)
 		if err != nil {
-			return fmt.Errorf(`validation for attribute %q failed during upstream refresh: %w`, attribute, err)
+			return nil, fmt.Errorf(`validation for attribute %q failed during upstream refresh: %w`, attribute, err)
 		}
 	}
-	// we checked that the user still exists and their information is the same, so just return.
-	return nil
-}
 
-func (p *Provider) performRefresh(ctx context.Context, userDN string) (*ldap.SearchResult, error) {
-	search := p.refreshUserSearchRequest(userDN)
-
-	conn, err := p.dial(ctx)
-	if err != nil {
-		return nil, fmt.Errorf(`error dialing host %q: %w`, p.c.Host, err)
+	// If we have group search configured, search for groups to update the value.
+	if len(p.c.GroupSearch.Base) > 0 {
+		mappedGroupNames, err := p.searchGroupsForUserDN(conn, userDN)
+		if err != nil {
+			return nil, err
+		}
+		sort.Strings(mappedGroupNames)
```

Comment:
Are we always sorting the results of searchGroupsForUserDN? If so, maybe the sort should move inside that function?